src/api/integrator/mapbox-gl.js

```diff
@@ -40,17 +41,18 @@ class MGLIntegrator {
     _unregisterMoveObserver(observerName) {
         delete this.moveObservers[observerName];
     }
-    addLayer(layerId, beforeLayerID, moveCallback, paintCallback) {
+    addLayer(layer, beforeLayerID) {
         const callbackID = `_cartoGL_${uid++}`;
-        this._registerMoveObserver(callbackID, moveCallback);
+        const layerId = layer.getId();
+        this._registerMoveObserver(callbackID, layer.requestData.bind(layer));
         this.map.repaint = true; // FIXME: add logic to manage repaint flag
         this.map.setCustomWebGLDrawCallback(layerId, (gl, invalidate) => {
-            if (!this.invalidateMGLWebGLState) {
-                this.invalidateMGLWebGLState = invalidate;
+            if (!this.invalidateWebGLState) {
+                this.invalidateWebGLState = invalidate;
                 this.notifyObservers();
                 this.renderer._initGL(gl);
             }
-            paintCallback();
+            layer.paintCallback();
```

Comment:
By removing the callbacks and passing the layer instead of just the layerID you are coupling the integrator to the Layer, but ok, maybe it's more clear this way

Reply:
Yes, the layer needs to match this interface now. But this code requires more work/refactor in the future.